aten/src/ATen/Allocator.h

```diff
@@ -3,14 +3,153 @@
 #include <memory>
 #include <stddef.h>
 
-#include "ATen/Retainable.h"
+#include <ATen/Error.h>
+#include <ATen/Retainable.h>
+#include <ATen/Device.h>
 
 namespace at {
 
+// Note [Supervisor deleter]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~
+// SupervisorPtr solves a common problem for allocators of tensor data, which
```

Comment:
I think it may be useful to split up this file into the actual API and the building blocks.  There are pretty big walls of text one has to get through before they even see what the actual API looks like (here, I'm considering the DevicePtr to be the actual API, and I realize some of this is due to having to define things in order in C++...)

Reply:
I.e. I view the SupervisorPtr/SupervisorPtr stuff as basically a different retainable-like abstraction.